<img src=cenapad.jpg align=left><BR><BR><BR><BR><BR><BR>
<h1>Curso Introducción al Lenguaje Python<br>Programación Estructurada</h1>
<img src=python-logo.png align=left>

# 2.1 Primeros pasos hacia la programación¶

Por supuesto, podemos usar Python para tareas más complicadas que sumar dos y dos. Por ejemplo, podemos escribir una subsecuencia inicial de la serie de Fibonacci así:

In [ ]:
# Series de Fibonacci:
# la suma de dos elementos define el siguiente
a, b = 0, 1
while b < 10:
    print(b)
    a, b = b, a+b

In [ ]:
lista=[]
a, b = 0, 1

while b < 10:
    lista.append(b)
    a, b = b, a+b
    
print(lista)

Este ejemplo introduce varias características nuevas.
<ul>
<li>La primer línea contiene una asignación múltiple: las variables a y b toman en forma simultanea los nuevos valores 0 y 1. En la última linea esto se vuelve a usar, demostrando que las expresiones a la derecha son evaluadas antes de que suceda cualquier asignación. Las expresiones a la derecha son evaluadas de izquierda a derecha.</li><br>

<li>El bucle while se ejecuta mientras la condición (aquí: b < 10) sea verdadera. En Python, como en C, cualquier entero distinto de cero es verdadero; cero es falso. La condición también puede ser una cadena de texto o una lista, de hecho cualquier secuencia; cualquier cosa con longitud distinta de cero es verdadero, las secuencias vacías son falsas. La prueba usada en el ejemplo es una comparación simple.</li><br>

<li>El cuerpo del bucle tiene sangría: Python utiliza 4 espacios para diferencias que se encuentra en un bucle o función. El interpretador permite la utilziación de 4 espacios o tabulación (tab) pero no mezclar ambos tipos, pues puede generar errores en la interpretación.</li><br>

<li>La función print() escribe el valor de el o los argumentos que se le pasan. Difiere de simplemente escribir la expresión que se quiere mostrar en la forma en que maneja múltiples argumentos, cantidades en punto flotante, y cadenas. Las cadenas de texto son impresas sin comillas, y un espacio en blanco es insertado entre los elementos, así puedes formatear cosas de una forma agradable:</li>

</ul>

In [ ]:
i = 256 * 256
print('El valor de i es', i)

El parámetro nombrado end puede usarse para evitar el salto de linea al final de la salida, o terminar la salida con una cadena diferente:

In [ ]:
lista = []
a, b = 0, 1
while b < 1000:
    print(b, end=', ')
    lista.append(b)
    a, b = b, a+b

In [ ]:
for elemento in lista:
    print(elemento, end=", ")

Otros ejemplos de la utilización de <b>print</b>.

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ')
print('Mars', 'Jupiter', 'Saturn', sep=', ')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ', end='\t\t')
print('Mars', 'Jupiter', 'Saturn', sep=', ', end='\t\t')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ')
print('Mars', 'Jupiter', 'Saturn', sep=', ')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Printing in a Nutshell', end='\n * ')
print('Calling Print', end='\n * ')
print('Separating Multiple Arguments', end='\n * ')
print('Preventing Line Breaks')

Para más información sobre print visitar: https://realpython.com/python-print/

# 2.2. Condicionales: if, elif y else

La declaración if/elif/else funciona para controlar el fluxo de ejecución basandose en una o más condiciones. La sintaxis es la siguiente:

In [ ]:
#No ejecute está célula
if condicion:
    comando1
    comando2
    ...
    comandoN
elif condicion:
    comando1
    comando2
    ...
    comandoN
else:
    comando1
    comando2
    ...
    comandoN

Observa la <b>IDENTACIÓN</b> que define los bloques de código, a diferencia de otros lenguajes que usan delimitadores como llaves tales como "{" y "}" o bloques con "begin" y "end".<br>

Una buena práctica es usar <b>IDENTATION</b> con 4 espacios (primer nivel), 8 espacios (segundo nivel), etc.<br>

<b>NO SE DEBÉN USAR TABULACIONES</b> (debido a las convenciones de portabilidad y legibilidad), aunque están permitidas.<br>

<b>NUNCA se debe mezclar espacios con tabulaciones en el mismo programa.</b><br>

IDENTATION en Python es obligatorio y forma parte de la sintaxis. La ventaja de esta característica es la legibilidad del código. La mayoría de los editores de código o IDE de desarrollo facilitarán el trabajo de IDENTACIÓN colocando espacios automáticamente si se encuentra dentro del contexto de un bloque de código.

#### Ejemplo de uso del condicional if elif else:

In [ ]:
nombre = input("¿Cual es su nombre?:")
edad = int(input("¿Cual es su edad?:")) #Ingrese un número entero
grupo=""

if edad <= 12:
    grupo = "Infancia"
elif edad <= 18:
    grupo = "Adolescencia"
elif edad <= 26:
    grupo = "Juventud"
elif edad <= 65:
    grupo = "Adultez"
else:
    grupo = "Persona Mayor"

print(f"Nombre: {nombre} - Edad: {edad} - Pertenece al grupo {grupo}.")

La palabra reservada ‘elif‘ es una abreviación de ‘else if’, y es útil para evitar un sangrado excesivo. Una secuencia if ... elif ... elif ... sustituye las sentencias switch o case encontradas en otros lenguajes.

La declaración if/elif/else puede precisa tener  tener ninguún bloque elif, por ejemplo:

In [ ]:
edad=21
if edad<=17:
    print("Menor de edad")
else:
    print("Mayor de edad")

Es posible combinar condiones con otros operadores <b>and</b>, <b>not</b> y <b>or</b>.

In [ ]:
edad = 31
if edad <= 12:
    print("Infancia")
elif edad > 12 and edad < 21:
    print("Adolescente")
elif edad>=21:
    print("Adulto")

In [1]:
x = int(input("Ingrese un número entero: "))
y = int(input("Ingrese otro número entero: "))
if x%2==1 and y%2==1:
    print("Ambos números son impares")
else:
    print("Por lo menos uno de los números es par")

Ingrese un número entero: 15
Ingrese otro número entero: 36
Por lo menos uno de los números es par


In [2]:
x = int(input("Ingrese un número entero: "))
y = int(input("Ingrese otro número entero: "))
if x%2==1 or y%2==1:
    print("Por lo menos uno delos números es impar")
else:
    print("Ambos números son pares")

Ingrese un número entero: 15
Ingrese otro número entero: 15
Por lo menos uno delos números es impar


In [3]:
num = list(range(1,30))
for n in num:
    if n%3==0:
        print(f"{n} es divisible por 3")

3 es divisible por 3
6 es divisible por 3
9 es divisible por 3
12 es divisible por 3
15 es divisible por 3
18 es divisible por 3
21 es divisible por 3
24 es divisible por 3
27 es divisible por 3


Alternativamente, en lugar de imprimir en la pantalla, podemos crear una lista con los números divisibles por 3 y luego imprimirla

In [4]:
num = list(range(1,30))
num_div3=[] #Iniciando una lista vacia (Esto es necesario para que podemos usar el método append)
for n in num:
    if n%3==0:
        num_div3.append(n)
print(num_div3)

[3, 6, 9, 12, 15, 18, 21, 24, 27]


### Ejercicio 1:

Dado tres puntos en el plano expresados como coordenadas (x, y) informar cuál es el que se encuentra más lejos del centro de coordenadas.

In [54]:
import math
listX = []
listY = []
x1 = int(input("X1 = "))
listX.append(x1)
y1 = int(input("Y1 = "))
listY.append(y1)
x2 = int(input("X2 = "))
listX.append(x2)
y2 = int(input("Y2 = "))
listY.append(y2)
x3 = int(input("X3 = "))
listX.append(x3)
y3 = int(input("Y3 = "))
listY.append(y3)

may = 0
punto = 0
i = 0
while i<3:
    dist = math.sqrt(listX[i]**2 + listY[i]**2)
    if(dist > may):
        punto = i
        may = dist
    i += 1
print(f"La distancia mas grande es la del punto ({listX[punto]},{listY[punto]})")
    


X1 = 1
Y1 = 1
X2 = 5
Y2 = 6
X3 = 2
Y3 = 4
La distancia mas grande es la del punto (5,6)


### Ejercicio 2:

Escribir un programa que almacene la cadena de caracteres contraseña en una variable, pregunte al usuario por la contraseña e imprima por pantalla si la contraseña introducida por el usuario coincide con la guardada en la variable sin tener en cuenta mayúsculas y minúsculas.

Sugerencia: utilizar la función lower.

<b>Salida deseada:</b><br>
La contaseña coincide<br>
<br>
La contraseña no coincide

In [10]:
pass1 = 'pass123'
pass2 = input("Password: ")
if pass1 == pass2.lower():
    print('La contrasena coincide')
else:
    print('La contrasena no coincide')


Password: pass123
La contrasena coincide


### Ejercicio 3:

Para tributar un determinado impuesto se debe ser mayor de 16 años y tener unos ingresos iguales o superiores a 2000 Bs mensuales. Escribir un programa que pregunte al usuario su edad y sus ingresos mensuales y muestre por pantalla si el usuario tiene que tributar o no.

<b>Salida deseada:</b><br>
Tienes que tributar<br>
<br>
No tienes que tributar

In [13]:
edad = int(input('Edad: '))
ingresos = int(input('Ingresos: '))
if edad > 16 and ingresos > 2000:
    print("Tienes que tributar")
else:
    print("No tienes que tributar")


Edad: 49
Ingresos: 6000
Tienes que tributar


### Ejercicio 4:

Escriba un programa que pida dos números enteros y que calcule su división, escribiendo si la división es exacta o no. El programa debe tomar en cuenta que no se puede dividir por cero.

In [15]:
num = int(input("Num: "))
div = int(input("Div: "))
if div == 0:
    print("No se puede dividir por 0")
elif num%div == 0:
    print("La division es exacta")
else:
    print("La division no es exacta")


Num: 5
Div: 2
La division no exacta


### Ejercicio 5:

Escriba un programa que pida año, mes y día y retorne cuantos días han pasado o faltan a la fecha actual.

In [30]:
from datetime import date
ano = int(input("Ano: "))
mes = int(input("Mes: "))
dia = int(input("Dia: "))
hoy=date.today()
fecha = date(ano,mes,dia)
if hoy>fecha:
    print(f'Pasaron {abs(hoy-fecha).days} dias')
else:
    print(f'Faltan {abs(fecha-hoy).days} dias')



Ano: 1998
Mes: 06
Dia: 9
Pasaron 7934 dias


### Ejercicio 6:

Escriba un programa que pida dos números enteros y que escriba si el mayor es múltiplo del menor. El programa debe avisar cuando se escriba valores negativos o nulos.

In [31]:
try:
    n1 = int(input("N1 = "))
    n2 = int(input("N2 = "))
except ValueError:
    print("Un valor es nulo")

mayor= max(n1, n2)
menor= min(n1, n2)
if n1 <= 0 or n2 <= 0:
    print("Uno de los numeros es menor a 0")
elif mayor%menor == 0:
    print("El mayor es multiplo del menor")
else:
    print("El mayor no es multiplo del menor")
    

N1 = 14
N2 = 7
El mayor es multiplo del menor


### Ejercicio 7:

Escriba un programa que pida los coeficientes de una ecuación de segundo grado (a x² + b x + c = 0) y escriba la solución.

Se recuerda que una ecuación de segundo grado puede no tener solución (solución con números complejos), tener una solución única, tener dos soluciones o que todos los números sean solución. Se recuerda que la fórmula de las soluciones cuando hay dos soluciones es x = (-b ± √(b2-4ac) ) / (2a)

En caso de no tener solución, hacer uso de la números complejos de Python.

Más información en:<br>
[1] https://relopezbriega.github.io/blog/2015/10/12/numeros-complejos-con-python/<br>
[2] https://pybonacci.org/2014/02/04/como-separar-parte-real-e-imaginaria-en-sympy/

<b>Salida deseada</b><br>
ECUACIÓN A X² + B X + C = 0<br>
Escriba el valor del coeficiente a: 2<br>
Escriba el valor del coeficiente b: -7<br>
Escriba el valor del coeficiente c: 3<br>
La ecuación tiene dos soluciones: 3.0 y 0.5

In [32]:
import math
print("ECUACIÓN A X² + B X + C = 0")
a = float(input("Escriba el valor del coeficiente a: "))
b = float(input("Escriba el valor del coeficiente b: "))
c = float(input("Escriba el valor del coeficiente c: "))
d = b**2 - 4*a*c
if d==0:
    sol = (-b)/(2*a)
    print(f"La ecuacion tiene una solucion: {sol:.2f}")
elif d>0:
    sol1= ((-b)+ math.sqrt(d))/(2*a)
    sol2= ((-b)- math.sqrt(d))/(2*a)
    print(f"La ecuacion tiene dos soluciones: {sol1:.2f} y {sol2:.2f}")
else: 
    com1 = complex(-b, math.sqrt(-d))
    com2 = complex(-b, -math.sqrt(-d))
    sol1 = (com1)/((2*a) +0j)
    sol2= (com2)/((2*a) +0j)
    print(f"La ecuacion tiene dos soluciones: {sol1:.2f} y {sol2:.2f}")



ECUACIÓN A X² + B X + C = 0
Escriba el valor del coeficiente a: 1
Escriba el valor del coeficiente b: 1
Escriba el valor del coeficiente c: 1
La ecuacion tiene dos soluciones: -0.50+0.87j y -0.50-0.87j


### Ejercicio 8:

Escriba un programa que pregunte primero si se quiere calcular el área de un triángulo o la de un círculo. Si se contesta que se quiere calcular el área de un triángulo (escribiendo T o t), el programa tiene que pedir entonces la base y la altura y escribir el área. Si se contesta que se quiere calcular el área de un círculo (escribiendo C o c), el programa tiene que pedir entonces el radio y escribir el área.

Se recuerda que el área de un triángulo es base por altura dividido por 2 y que el área de un círculo es Pi (aproximadamente 3,141592) por el radio al cuadrado. Utilizar 2 decimales para la salida.

Nota: Utilice la libreria math para obtener la función pi.

<b>Salida deseada:</b><br>
CÁLCULO DE ÁREAS<br>
Elija una figura geométrica:<br>
a) Triángulo<br>
b) Círculo<br>
¿Qué figura quiere calcular (Escriba T o C)? T<br>
Escriba la base: 3<br>
Escriba la altura: 5.5<br>
Un triángulo de base 3.0 y altura 5.0 tiene un área de 8.25<br>
<br>
CÁLCULO DE ÁREAS<br>
Elija una figura geométrica:<br>
a) Triángulo<br>
b) Círculo<br>
¿Qué figura quiere calcular (Escriba T o C)? C<br>
Escriba el radio: 2<br>
Un círculo de radio 2.0 tiene un área de 12.57<br>

In [1]:
import math
print("""CÁLCULO DE ÁREAS 
Elija una figura geométrica:
a) Triángulo
b) Círculo """)
figura = input("¿Qué figura quiere calcular (Escriba T o C)?")
if figura.lower() == "t":
    base = float(input("Escriba la base: "))
    alto = float(input("Escriba la altura: "))
    area = (base*alto) /2
    print(f'Un triangulo de base {base:.2f} y altura {alto:.2f} tiene un area de {area:.2f}')
elif figura.lower() == "c":
    radio = float(input("Escriba el radio: "))
    area = radio**2 * math.pi
    print(f'Un circulo de radio {radio:.2f} tiene un area de {area:.2f}')
else:
    print("Respuesta incorrecta")


CÁLCULO DE ÁREAS 
Elija una figura geométrica:
a) Triángulo
b) Círculo 
¿Qué figura quiere calcular (Escriba T o C)?c
Escriba el radio: 2
Un circulo de radio 2.00 tiene un area de 12.57


# 2.3. La Sentencia for

La sentencia <b>for</b> en Python difiere un poco de lo que uno puede estar acostumbrado en lenguajes como C o Pascal. En lugar de siempre iterar sobre una progresión aritmética de números (como en Pascal) o darle al usuario la posibilidad de definir tanto el paso de la iteración como la condición de fin (como en C), la sentencia <b>for</b> de Python itera sobre los ítems de cualquier secuencia (una lista o una cadena de texto), en el orden que aparecen en la secuencia. Por ejemplo:

In [1]:
# Midiendo cadenas de texto
palabras = ['gato', 'ventana', 'defenestrado']
for p in palabras:
    print(p, len(p))

gato 4
ventana 7
defenestrado 12


Si necesitá modificar la secuencia sobre la que estás iterando mientras estás adentro del ciclo (por ejemplo para borrar algunos ítems), se recomienda que hagas primero una copia. Iterar sobre una secuencia no hace implícitamente una copia. La notación abajo es especialmente conveniente para esto:

In [ ]:
for p in palabras[:]:  # hace una copiade toda la lista
    if len(p) > 6:
        palabras.insert(0, p)

print(palabras)

Con for w in words:, el ejemplo intentaría crear una lista infinita, insertando defenestrado una y otra vez.<br>

Un bucle <b>for</b> funciona como un iterador. Un <b>for</b> se utiliza para realizar acciones sobre elementos de una secuencia o cualquier elemento iterable, como cadenas, listas y otros objetos iterables que veremos más adelante.

<b>Formato general para una declaración for</b>

for item in object:<br>
&nbsp;&nbsp;&nbsp;&nbsp;statements to do stuff

También es común usar un rango como iterable en un <b>for</b>

for i in range(0,10):<br>
&nbsp;&nbsp;&nbsp;&nbsp;hace alguna cosa con i

#### Ejemplo. El complemento de una secuencia de ADN se obtiene reemplazando todo As con Ts, Ts con As, Gs con Cs y Cs con Gs por Gs

In [2]:
dna = "AGCAGGTGCGAGTAGCAGTAGCA"
dna

'AGCAGGTGCGAGTAGCAGTAGCA'

In [3]:
#Transformar todas las A's en Ts, Ts en As, Gs en Cs y Cs en Gs
complemento=""  #Iniciando string vacia
for x in dna:  #A cada iteración en este bucle, x será una letra del dna
    if x=="G":  #Si x fuera G, concatena la string complemento con un C
        complemento = complemento + "C"
    elif x=="C": #Si x fuera C, concatena la string complemento con un T
        complemento = complemento + "G"
    elif x == "A":
        complemento = complemento + "T"
    elif x == "T":
        complemento = complemento + "A"
        
print(complemento)

TCGTCCACGCTCATCGTCATCGT


<b>join</b> es un método de la clase string que concatena una cadena con las cadenas de una lista.<br><br>
<b>reversed</b> devuelve una lista o cadena invertida.<br><br>
Vamos a invertir una string y usar el método de join para formar una nueva cadena con el inverso complementario de la secuencia de DNA.

In [7]:
reverso_complementar = ''.join(reversed(complemento)) #conatena la string vacia '' con el complemento invertido
print(f"DNA={dna}\nREVCOMP={reverso_complementar}")
print(complemento[::-1])

DNA=AGCAGGTGCGAGTAGCAGTAGCA
REVCOMP=TGCTACTGCTACTCGCACCTGCT
TGCTACTGCTACTCGCACCTGCT


In [8]:
for i in range(len(dna)):
    print(f"Caracter de índice {i} é um {dna[i]}")

Caracter de índice 0 é um A
Caracter de índice 1 é um G
Caracter de índice 2 é um C
Caracter de índice 3 é um A
Caracter de índice 4 é um G
Caracter de índice 5 é um G
Caracter de índice 6 é um T
Caracter de índice 7 é um G
Caracter de índice 8 é um C
Caracter de índice 9 é um G
Caracter de índice 10 é um A
Caracter de índice 11 é um G
Caracter de índice 12 é um T
Caracter de índice 13 é um A
Caracter de índice 14 é um G
Caracter de índice 15 é um C
Caracter de índice 16 é um A
Caracter de índice 17 é um G
Caracter de índice 18 é um T
Caracter de índice 19 é um A
Caracter de índice 20 é um G
Caracter de índice 21 é um C
Caracter de índice 22 é um A


In [9]:
#El ciclo for anterior también se puede hacer usando la función enumerate
for i,x in enumerate(dna):
    print(f"Caracter de índice {i} é um {x}")

Caracter de índice 0 é um A
Caracter de índice 1 é um G
Caracter de índice 2 é um C
Caracter de índice 3 é um A
Caracter de índice 4 é um G
Caracter de índice 5 é um G
Caracter de índice 6 é um T
Caracter de índice 7 é um G
Caracter de índice 8 é um C
Caracter de índice 9 é um G
Caracter de índice 10 é um A
Caracter de índice 11 é um G
Caracter de índice 12 é um T
Caracter de índice 13 é um A
Caracter de índice 14 é um G
Caracter de índice 15 é um C
Caracter de índice 16 é um A
Caracter de índice 17 é um G
Caracter de índice 18 é um T
Caracter de índice 19 é um A
Caracter de índice 20 é um G
Caracter de índice 21 é um C
Caracter de índice 22 é um A


### Ejercicio 9:

Escriba un programa que pida un número de dados y muestre los valores de ese número de dados, al azar.

Nota. Utilizar la libreria random.
Más información en:
[1] https://stackoverflow.com/questions/3996904/generate-random-integers-between-0-and-9

<b>Salida deseada:</b><br><br>
TIRADA DE DADOS<br>
Número de dados: 0<br>
¡Imposible!<br>
<br>
TIRADA DE DADOS<br>
Número de dados:  5<br>
Dados: 3 5 1 1 3<br>

In [13]:
from random import randrange
print("TIRADA DE DADOS")
numDados = int(input("Numero de dados: "))
if numDados <= 0:
    print("imposible")
else:
    print("Dados: ")
    for i in range(0,numDados):
        print(randrange(1,7), end = " ")

TIRADA DE DADOS
Numero de dados: 6
Dados: 
5 5 1 4 6 3 

### Ejercicio 10:

Escriba un programa que pida un número de jugadores y tire un dado para cada jugador. El jugador que tire el mayor dado ganará el juego.

<b>Salida deseada:</b><br><br>
TIRADAS DE DADO<br>
Número de jugadores: 1<br>
¡Imposible!<br>
<br>
TIRADAS DE DADO<br>
Número de jugadores: 4<br>
Jugador 1: 3<br>
Jugador 2: 3<br>
Jugador 3: 5<br>
Jugador 4: 1<br>
El ganador es el Jugador 3.<br>
<br>
TIRADAS DE DADO<br>
Número de jugadores: 3<br>
Jugador 1: 4<br>
Jugador 2: 4<br>
Jugador 3: 1<br>
Los ganadores son el Jugador 1, Jugador 2.

In [38]:
from random import randrange
print("TIRADAs DE DADOS")
numJug = int(input("Numero de jugadores: "))
dados=[]
if numJug <= 1:
    print("¡Imposible!")
else:
    for i in range(0,numJug):
        dadoJ = randrange(1,7)
        dados.append(dadoJ)
        print(f"Jugador {i+1}: {dadoJ}" )
    mayorDado = max(dados)
    ganadores = []
    for i, x in enumerate(dados):
        if x == mayorDado:
            ganadores.append(i)
    if len(ganadores) > 1:
        print("Los ganadores son el ", end = "")
        for i in ganadores:
            if (len(ganadores)-1) == i:
                print(f"Jugador {i+1}", end=". ")
            else:
                print(f"Jugador {i+1}", end=", ")
    else:
        print(f"El ganador es el Jugador {ganadores[0]+1}")
            
            
            


TIRADAs DE DADOS
Numero de jugadores: 2
Jugador 1: 2
Jugador 2: 4
El ganador es el Jugador 2


### Ejercicio 11:

Escriba un programa que pida un número de dados y tire esa cantidad de dados. El primer jugador obtiene un punto por cada dado par. El segundo jugador obtiene un punto por cada dado impar. El jugador que saque más puntos, gana.

<b>Salida deseada:</b><br><br>
PARES Y NONES<br>
Número de dados: 0<br>
¡Imposible!<br>
<br>
PARES Y NONES<br>
Número de dados: 7<br>
Dados: 4 5 2 3 2 1 6<br>
Ha ganado el jugador de los pares.<br>
<br>
PARES Y NONES<br>
Número de dados: 8<br>
Dados: 1 4 1 3 5 6 6 2<br>
Han empatado.

In [32]:
from random import randrange
print("PARES Y NONES")
numDados = int(input("Numero de dados: "))
if numDados <= 0:
    print("¡Imposible!")
else:
    puntajeJ1 = 0
    puntajeJ2 = 0
    print("Dados: " , end = "")
    for i in range(0,numDados):
        dado = randrange(1,7)
        print(dado, end = " ")
        if dado%2 == 0:
            puntajeJ1 += 1
        else:
            puntajeJ2 += 1
    if puntajeJ1 > puntajeJ2:
        print("\nHa ganado el jugador de los pares.")
    else:
        print("\nHa ganado el jugador de los impares.")
        
        
    

PARES Y NONES
Numero de jugadores: 10
Dados: 1 6 6 2 6 4 5 5 2 2 
 Ha ganado el jugador de los pares.


### Ejercicio 12:

Escriba un programa que pida un número de dados y tire esa cantidad de dados. Si no salen dos dados iguales seguidos, el jugador gana. Si salen, pierde.

<b>Salida deseada:</b><br><br>
DADOS IGUALES<br>
Número de dados: 1<br>
¡Imposible!<br>
<br>
DADOS IGUALES<br>
Número de dados: 6<br>
Dados: 3 1 1 4 6 2<br>
El jugador ha perdido.<br>
<br>
DADOS IGUALES<br>
Número de dados: 4<br>
Dados: 4 2 5 3<br>
El jugador ha ganado.<br>

In [42]:
from random import randrange
print("DADOS IGUALES")
numDados = int(input("Numero de dados: "))
if numDados <= 1:
    print("¡Imposible!")
else:
    print("Dados: " , end = "")
    ganar = True
    prevD = 7
    for i in range(0,numDados):
        dado = randrange(1,7)
        print(dado, end = " ")
        if dado == prevD:
            ganar = False
        else:
            prevD = dado
    if ganar:
        print("\nEl jugador ha ganado.")
    else:
        print("\nEl jugador ha perdido.")
        

DADOS IGUALES
Numero de dados: 3
Dados: 3 1 4 
El jugador ha ganado.


### Ejercicio 13:

Escribir un programa que pregunte al usuario una cantidad a invertir, el interés anual y el número de años, y muestre por pantalla el capital obtenido en la inversión cada año que dura la inversión.

<b>Salida deseada:</b><br><br>
¿Cantidad a invertir?  1000<br>
¿Interés porcentual anual?  10<br>
¿Años? 5<br>
Capital tras 1 años: 1100.00<br>
Capital tras 2 años: 1210.00<br>
Capital tras 3 años: 1331.00<br>
Capital tras 4 años: 1464.10<br>
Capital tras 5 años: 1610.51<br>

In [44]:
invertir = float(input("¿Cantidad a invertir?"))
interes = float(input("¿Interés porcentual anual?"))
anos = int(input("¿Anos?"))
if invertir < 0 or interes < 0 or anos < 0:
    print("Imposible")
else:
    interes = interes/100
    for i in range(0,anos):
        invertir = invertir + invertir*interes
        print(f"Capital tras {i+1} anos: {invertir:.2f}")

¿Cantidad a invertir?200
¿Interés porcentual anual?5
¿Anos?7
Capital tras 1 anos: 210.00
Capital tras 2 anos: 220.50
Capital tras 3 anos: 231.53
Capital tras 4 anos: 243.10
Capital tras 5 anos: 255.26
Capital tras 6 anos: 268.02
Capital tras 7 anos: 281.42


# Ejercicio 14:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo ("* ") como el de más abajo, de altura el número introducido.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo) 5<br>
*<br>
\* *<br>
\* * *<br>
\* * * *<br>
\* * * * * 

In [45]:
puntos = int(input("Introduce la altura del triángulo (entero positivo)"))
if puntos <= 0:
    print("Imposible")
else:
    for i in range(0,puntos):
        print("")
        for y in range(0,i+1):
            print("*", end = "")

Introduce la altura del triángulo (entero positivo)5

*
**
***
****
*****

### Ejercicio 15:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo como el de más abajo.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  9<br>

1 <br>
3 1 <br>
5 3 1 <br>
7 5 3 1 <br>
9 7 5 3 1

In [19]:
altura = int(input("Introduce la altura del triángulo (entero positivo): "))
if altura <= 0:
    print("Imposible")
else:
    lista = list(range(1,altura+1))[::2]
    for i in range(0,len(lista)):
        for y in range(i,-1,-1):
            print(lista[y], end = " ")
        print("")

Introduce la altura del triángulo (entero positivo): 6
1 
3 1 
5 3 1 


### Ejercicio 16:

Escriba un programa que pida un número entero mayor que cero y que escriba sus divisores.

<b>Salida deseada:</b><br><br>
DIVISORES<br>
Escriba un número mayor que cero: -5<br>
¡Le he pedido un número entero mayor que cero!<br>
<br>
DIVISORES<br>
Escriba un número entero mayor que cero: 200<br>
Los divisores de 200 son 1 2 4 5 8 10 20 25 40 50 100 200

In [21]:
print("Divisores")
num = int(input("Escriba un número mayor que cero: "))
if num <= 0:
    print("¡Le he pedido un número entero mayor que cero!")
else:
    print(f"Los divisores de {num} son:", end = " ")
    for i in range(1,num+1):
        if num % i == 0:
            print(i, end=" ")
            


Divisores
Escriba un número mayor que cero: 521
Los divisores de 521 son: 1 521 

### Ejercicio 17:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo como el de número de primos.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  20<br>

1 <br>
3 1 <br>
5 3 1 <br>
7 5 3 1 <br>
11 7 5 3 1 <br>
13 11 7 5 3 1 <br>
17 13 11 7 5 3 1 <br>
19 17 13 11 7 5 3 1

In [25]:
altura = int(input("Introduce la altura del triángulo (entero positivo): "))
if altura <= 0:
    print("Imposible")
else:
    primos= [2]
    for i in range(3,altura+1):
        primo = True
        for y in range(2,i):
            if i % y == 0:
                primo = False
                break
        if primo:
            primos.append(i)
    for i in range(0,len(primos)):
        for y in range(i,-1,-1):
            print(primos[y], end = " ")
        print("")
    

Introduce la altura del triángulo (entero positivo): 25
2 
3 2 
5 3 2 
7 5 3 2 
11 7 5 3 2 
13 11 7 5 3 2 
17 13 11 7 5 3 2 
19 17 13 11 7 5 3 2 
23 19 17 13 11 7 5 3 2 


### Ejercicio 18:

Escriba un programa que pida dos números enteros y escriba la suma de todos los enteros desde el primer número hasta el segundo.

<b>Salida deseada:</b><br><br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: -5<br>
¡Le he pedido un número entero positivo!<br>
<br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: 7<br>
Escriba un número entero mayor que 7: 7<br>
¡Le he pedido un número entero mayor que 7!<br>
<br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: 30<br>
Escriba un número entero mayor que 30: 32<br>
30 + 31 + 32 = 93<br>

In [28]:
num1 = int(input("Escriba un número entero positivo: "))
if num1 <= 0:
    print("¡Le he pedido un número entero positivo!")
else:
    num2 = int(input(f"Escriba un número entero mayor que {num1}: "))
    if num2 < num1:
        print(f"¡Le he pedido un número entero mayor que {num1}!")
    else:
        suma = 0
        for i in range(num1,num2+1):
            suma += i
            if i == num2:
                print(f"{i} = ", end = " ")
            else:
                print(f"{i} + ", end = " ")
        
        print(suma)

Escriba un número entero positivo: -1
¡Le he pedido un número entero positivo!


### Ejercicio 19:

Escriba un programa que pida un número entero mayor que cero y calcule su factorial.

<b>Salida deseada:</b><br><br>
FACTORIAL<br>
Escriba un número entero mayor que cero: -5<br>
¡Le he pedido un número entero mayor que cero!<br>
<br>
FACTORIAL<br>
Escriba un número entero mayor que cero: 5<br>
El factorial de 5 es 120<br>

In [30]:
num1 = int(input("Escriba un número entero  mayor que cero: "))
if num1 <= 0:
    print("¡Le he pedido un número entero mayor que cero!")
else:
    factorial = 1
    for i in range(1,num1+1):
        factorial*= i
    print(f"El factorial de {num1} es {factorial}")

Escriba un número entero  mayor que cero: 1
El factorial de 1 es 1


### Ejercicio 20:

Escribir un programa que pida al usuario un número entero (del 1 al 10) y genere la matriz caracol.

<b>Salida deseada:</b><br><br>

Introduce un número (entero positivo):  3<br>

1 2 3<br>
8 9 4<br>
7 6 5<br>

In [31]:

import numpy
num=int(input("Introduce un número (entero positivo): "))
if num>0 and num<11:
    mat = numpy.zeros(shape=(num,num))
    cont=1;
    iY1=1
    iX1=num-1
    iY2=num-1
    iX2=num-2
    iY3=num-2
    iX3=0
    iY4=1
    iX4=1
    for i in range(0,num):
        mat[0][i]=cont
        cont+=1
    for i in range(num-1,0,-2):
        for j in range(0,i):
            if mat[iY1+j][iX1]!=0:
                break
            mat[iY1+j][iX1]=cont
            cont+=1
        for j in range(0,i):
            if mat[iY2][iX2-j]!=0:
                break
            mat[iY2][iX2-j]=cont
            cont+=1
        iY1+=1
        iX1-=1
        iY2-=1
        iX2-=1
        i-=1
        for j in range(0,i):
            if mat[iY3-j][iX3]!=0:
                break
            mat[iY3-j][iX3]=cont
            cont+=1
        for j in range(0,i):
            if mat[iY4][iX4+j]!=0:
                break
            mat[iY4][iX4+j]=cont
            cont+=1
        iY3-=1
        iX3+=1
        iY4+=1
        iX4+=1
    print(mat)
else:
    print("Debe ingresar un entero del 1 al 10")    

Introduce un número (entero positivo): 7
[[ 1.  2.  3.  4.  5.  6.  7.]
 [24. 25. 26. 27. 28. 29.  8.]
 [23. 40. 41. 42. 43. 30.  9.]
 [22. 39. 48. 49. 44. 31. 10.]
 [21. 38. 47. 46. 45. 32. 11.]
 [20. 37. 36. 35. 34. 33. 12.]
 [19. 18. 17. 16. 15. 14. 13.]]


# 2.4. La Sentencia while

Una instrucción while prueba una condición inicial y comienza a ejecutar los comandos del bloque si la condición es verdadera. Mientras la condición permanezca verdadera, los comandos de bucle se ejecutan repetidamente.

Sintaxis:

mientras condición:

     comandos

#### Ejemplo de declaración while

In [ ]:
i=0
while i < 10:
    i+=1
    print(i)

In [3]:
# Iniciando con una lista vacia
nombres = []

# Defina una string llamada nuevo_nombre con cualquier valor diferente de "salir"
nuevo_nombre = ''

El ciclo while se ejecutará mientras el usuario ingrese nombres distintos a "salir", almacenando cada nombre ingresado en una lista.

In [ ]:
while nuevo_nombre != 'salir':
    # Pida que el usuario ingrese un nombre
    nuevo_nombre = input("Ingrese un nombre o la palabra salir: ")
    # Adiciona el nombre al final de la lista de nombres
    nombres.append(nuevo_nombre)

#Extraiga la palabra salir de la lista de names ("salir")
ultima = nombres.pop()
# Imprime la lista de nombres
print(nombres)

El control de bucles break, continue y pass permite:

In [ ]:
while True:
    line = input('Input > ')
    if line == 'done':
        break
    print(line)
print('Done!')

In [13]:
while True:
    line = input('Input > ')
    if line[0] == '#':
        continue
    if line == 'done':
        break
    print(line)
print('Done!')

Input > dsfdsf
dsfdsf
Input > dsf
dsf
Input > done
Done!


Más información:<br>
[1] https://www.pythonmania.net/es/2013/04/05/control-de-bucles-break-continue-y-pass/

### Ejercicio 21:

Escriba un programa que pida dos números enteros. El programa pedirá de nuevo el segundo número mientras no sea mayor que el primero. El programa terminará escribiendo los dos números.

<b>Salida deseada:</b><br><br>
NÚMERO MAYOR<br>
Escriba un número: 6<br>
Escriba un número mayor que 6: 6<br>
6 no es mayor que 6. Inténtelo de nuevo: 1<br>
1 no es mayor que 6. Inténtelo de nuevo: 8<br>
<br>
Los números que ha escrito son 6 y 8.

In [6]:
print("NÚMERO MAYOR")
num = int(input("Escriba un número: "))
num2 = int(input(f"Escriba un número mayor que {num}: "))
while num2 <= num:
    num2 = int(input(f"{num2} no es mayor que {num}. Inténtelo de nuevo:" ))
print(f"Los números que ha escrito son {num} y {num2}")

NÚMERO MAYOR
Escriba un número: 6
Escriba un número mayor que 6: 6
6 no es mayor que 6. Inténtelo de nuevo:5
5 no es mayor que 6. Inténtelo de nuevo:2
2 no es mayor que 6. Inténtelo de nuevo:3
3 no es mayor que 6. Inténtelo de nuevo:8
Los números que ha escrito son 6 y 8


### Ejercicio 22:

Escriba un programa que pida números decimales mientras el usuario escriba número mayores que el anterior.


<b>Salida deseada:</b><br><br>
NÚMEROS MAYORES<br>
Escriba un número: 7<br>
Escriba un número mayor que 7.0: 1<br>
1.0 no es mayor que 7.0.<br>
<br>
NÚMEROS MAYORES<br>
Escriba un número: 9.3<br>
Escriba un número mayor que 9.3: 9.3<br>
9.3 no es mayor que 9.3.<br>
<br>
NÚMEROS MAYORES<br>
Escriba un número: 4.5<br>
Escriba un número mayor que 4.5: 5<br>
Escriba otro número mayor que 5.0: 7.3<br>
Escriba otro número mayor que 5.0: 2<br>
2.0 no es mayor que 4.5.

In [8]:
print("NÚMEROS MAYORES")
num = float(input("Escriba un número: "))
num2 = float(input(f"Escriba un número mayor que {num}: "))
while num2 > num:
    num = num2
    num2 = float(input(f"Escriba otro número mayor que {num2}:"))
print(f"{num2} no es mayor que {num}")

NÚMEROS MAYORES
Escriba un número: 4.5
Escriba un número mayor que 4.5: 5
Escriba otro número mayor que 5.0:6
Escriba otro número mayor que 6.0:8
Escriba otro número mayor que 8.0:4
4.0 no es mayor que 8.0


### Ejercicio 23:

Escriba un programa que pida un valor límite positivo y a continuación pida números hasta que la suma de los números introducidos supere el límite inicial.


<b>Salida deseada:</b><br><br>
HASTA EL LÍMITE<br>
Escriba el valor límite: -3<br>
El límite debe ser mayor que 0. Inténtelo de nuevo: 0<br>
El límite debe ser mayor que 0. Inténtelo de nuevo: 5.5<br>
Escriba un número: 7.2<br>
Ha superado el límite. La suma de los números introducidos es 7.2.<br>
<br>
HASTA EL LÍMITE<br>
Escriba el valor límite: 12<br>
Escriba un número: 5.2<br>
Escriba otro número: -1<br>
Escriba otro número: 9<br>
Ha superado el límite. La suma de los números introducidos es 13.2.

In [12]:
print("HASTA EL LÍMITE")
num = float(input("Escriba el valor límite:"))
while num <= 0:
    num = float(input(f"El límite debe ser mayor que 0. Inténtelo de nuevo:"))
suma = 0
while suma < num:
    num2 = float(input(f"Escriba un numero: "))
    suma += num2
print(f"Ha superado el límite. La suma de los números introducidos es {suma}.")


HASTA EL LÍMITE
Escriba el valor límite:6
Escriba un numero: 2
Escriba un numero: 3
Escriba un numero: 4
Ha superado el límite. La suma de los números introducidos es 9.0.


### Ejercicio 24:

Escriba un programa que calcule la descomposición en factores primos de un número.

<b>Salida deseada:</b><br><br>
DESCOMPOSICIÓN EN FACTORES PRIMOS<br>
Escriba un número entero mayor que 1: 500<br>
Descomposición en factores primos: 2 2 5 5 5<br>
<br>
DESCOMPOSICIÓN EN NÚMEROS PRIMOS<br>
Escriba un número entero mayor que 1: 521<br>
Descomposición en factores primos: 521<br>
<br>
DESCOMPOSICIÓN EN NÚMEROS PRIMOS<br>
Escriba un número entero mayor que 1: 1<br>
1 no es mayor que 1. Inténtelo de nuevo: 720<br>
Descomposición en factores primos: 2 2 2 2 3 3 5

In [9]:
print("DESCOMPOSICIÓN EN FACTORES PRIMOS")
num = int(input("Escriba un número entero mayor que 1:"))
while num <= 1:
    num = int(input(f"{num} no es mayor que 1. Inténtelo de nuevo: "))
primo = 2
print("Descomposicion de factores primos: ", end="")
while num > 1:
    if num%primo == 0:
        num //= primo
        print(primo, end=" ")
    else:
        for i in range(primo+1,num+1):
            esPrimo = True
            for y in range(2,i):
                if i%y==0:
                    esPrimo = False
                    break
            if esPrimo:
                primo = i
                break
        
        
        
    
    


DESCOMPOSICIÓN EN FACTORES PRIMOS
Escriba un número entero mayor que 1:720
Descomposicion de factores primos: 2 2 2 2 3 3 5 

### Ejercicio 25:

Escriba un programa que muestre números al azar del 1 al 6 mientras lo pida el usuario. El programa debe mostrar la suma de los números anteriores.

<b>Salida deseada:</b><br>

Juego del Azar<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 2 - Suma: 2<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 5 - Suma: 7<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 4 - Suma: 11<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: x<br>
Programa terminado.

In [5]:
from random import randrange
print("Juego del Azar")
st = input("Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:")
suma = 0
while st == "":
    dado = randrange(1,7)
    suma += dado
    print(f"Tirada: {dado} - Suma: {suma}")
    st = input("Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:")
print("Programa Terminado")
    



Juego del Azar
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:
Tirada: 4 - Suma: 4
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:
Tirada: 2 - Suma: 6
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:
Tirada: 1 - Suma: 7
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:
Tirada: 1 - Suma: 8
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:l
Programa Terminado


### Ejercicio 26:

Escriba un programa que pida un número del 1 al 6, un número de jugadores y tire un dado para cada jugador. El 1er jugador obtenga el mismo número será el ganador.

<b>Salida deseada:</b><br>

JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 1<br>
¡Imposible!<br>
<br>
JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 4<br>
Jugador 1: 3<br>
Jugador 2: 3<br>
Jugador 3: 5<br>
El ganador es el Jugador 3.<br>
<br>
JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 3<br>
Jugador 1: 4<br>
Jugador 2: 4<br>
Jugador 3: 1<br>
No existe ganadores.

In [17]:
from random import randrange
print("Juego del Azar")
acertar = int(input("Número a acertar:"))
jugadores = int(input("Número de jugadores: "))
if acertar > 6 or acertar < 1 or jugadores < 2:
    print("Imposible")
else:
    ganador = True
    while ganador:
        for i in range(0,jugadores):
            dado = randrange(1,7)
            print(f"Jugador {i+1}: {dado}")
            if dado == acertar:
                print(f"El ganador es el Jugador {i+1}")
                ganador = False
                break

        

Juego del Azar
Número a acertar:5
Número de jugadores: 3
Jugador 1: 4
Jugador 2: 6
Jugador 3: 6
Jugador 1: 2
Jugador 2: 4
Jugador 3: 1
Jugador 1: 6
Jugador 2: 2
Jugador 3: 6
Jugador 1: 4
Jugador 2: 6
Jugador 3: 6
Jugador 1: 5
El ganador es el Jugador 1


### Ejercicio 27:

Escribir un programa que genere <b>una cadena de 4 caracteres aleatorios</b> que usen minúsculas y números en una variable, pregunte al usuario si desea averiguar la cadena o utilizar fuerza bruta para conseguir la contraseña correcta. En caso de usar la opción de fuerza bruta, declarar la cantidad de iteraciones utilizadas para encontrar la contraseña por fuerza bruta.

<b>Salida deseada:</b><br>

QUEBRAR UNA CONTRASEÑA<br>

El programa generó una contraseña.<br>

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  1234<br>
¡Incorrecto!

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  pass<br>
¡Contraseña correcta!<br>
<br>
<br>
QUEBRAR UNA CONTRASEÑA<br>

El programa generó una contraseña.<br>

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  cont<br>
¡Incorrecto!

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: fb<br>
Se realizó 3465 iteraciones para encontrar la contraseña correcta: abc5<br>


In [23]:
import string
import random
print("QUEBRAR UNA CONTRASEÑA")
lett = string.ascii_lowercase + string.digits
print(lett)
cadena = ""
for i in range(0,4):
    cadena += random.choice(lett)
noAdivinado = True
def makeString():
    cont = 0
    for i in range(0, len(lett)):
        a = lett[i]
        for y in range(0, len(lett)):
            b = lett[y]
            for z in range(0, len(lett)):
                c = lett[z]
                for w in range(0, len(lett)):
                    cont += 1
                    d = lett[w]
                    intento = a+b+c+d
                    if intento == cadena:
                        print(f"Se realizó {cont} iteraciones para encontrar la contraseña correcta: {intento}")
                        return 
     

print("El programa generó una contraseña.")
print(cadena)
while noAdivinado: 

    opcion = input("¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?:")
    if opcion == "a":
        guess = input("Introduce la contraseña: ")
        if guess == cadena:
            print("Correcto")
            noAdivinado = False
        else:
            print("Incorrecto")
    elif opcion == "fb":
        makeString()
        noAdivinado = False
    else:
        print("Elige una opcion")



QUEBRAR UNA CONTRASEÑA
abcdefghijklmnopqrstuvwxyz0123456789
El programa generó una contraseña.
o5cu
¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?:fb
Se realizó 693453 iteraciones para encontrar la contraseña correcta: o5cu


### Ejercicio 28:

Escriba un programa para jugar a adivinar un número (el ordenador "piensa" el número y el usuario tiene que adivinarlo). El programa empieza pidiendo entre qué números está el número a adivinar, se "inventa" un número al azar y después el usuario va probando valores y el programa va diciendo si son demasiado grandes o pequeños.

<b>Salida deseada:</b><br>

Valor mínimo: 0<br>
Valor máximo: 100<br>
Intente adivinar un número entero entre 0 y 100.<br>
Escriba un número: 20<br>
¡Demasiado pequeño! Inténtelo de nuevo: 30<br>
¡Demasiado grande! Inténtelo de nuevo: 27<br>
¡Acertó! Le ha costado 3 intentos.

<b>Sugerencia:</b> puede perfeccionar este programa haciendo:

* que al principio el programa se asegure de que el valor máximo es superior al valor mínimo.
* que el programa detecte "trampas", por ejemplo, si al decir 25 le decimos "mayor" y al decir 26 le decimos "menor", el programa debe decirnos que estamos haciendo trampas y dejar de jugar.

In [25]:
from random import randrange
minimo = int(input("Valor minimo: "))
maximo = int(input("Valor maximo: "))
print(f"Intente adivinar un numero entero entre {minimo} y {maximo}")
adiv = randrange(minimo,maximo+1)
num = int(input("Escriba un numero: "))
cont = 1
while num != adiv:
    if num > adiv:
        num = int(input("¡Demasiado grande! Inténtelo de nuevo: "))
    else:
        num = int(input("¡Demasiado pequeño! Inténtelo de nuevo: "))
    cont += 1
print(f"¡Acertó! Le ha costado {cont} intentos.")


Valor minimo: 0
Valor maximo: 10
Intente adivinar un numero entero entre 0 y 10
Escriba un numero: 5
¡Acertó! Le ha costado 1 intentos.


### Ejercicio 29:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo de Fibonacci.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  7<br>

1 <br>
1 1<br>
1 1 2<br>
1 1 2 3<br>
1 1 2 3 5<br>
1 1 2 3 5 8<br>
1 1 2 3 5 8 13<br>

In [40]:
altura = int(input("Introduce la altura del triángulo (entero positivo): "))
def fibo(n):
    if n <= 2:
        return 1
    else:
        return fibo(n-1) + fibo(n-2)

for i in range(0,altura+1):
    for y in range(0,i):
        num = fibo(y+1)
        print(num, end = " ")
    print("")


Introduce la altura del triángulo (entero positivo): 7

1 
1 1 
1 1 2 
1 1 2 3 
1 1 2 3 5 
1 1 2 3 5 8 
1 1 2 3 5 8 13 


### Ejercicio 30:

Implemente un programa que sea capaz de calcular el resultado de aplicarla formula siguiente (n i)=n!/(i!\*(n-i)!). Para ello usted debe leer un numero por teclado que debe ser mayor e igual que cero teniendo en cuenta además que n tiene que ser necesariamente mayor que i y además calcular el factorial de un número.

In [6]:
n = int(input("n: "))
i = int(input("i: "))
def factorial(p):
    fact = 1
    for i in range(1,p+1):
        fact*= i
    print(fact)
    return fact
if n < 0 or n < i:
    print("Datos incorrectos")
else:
    op = (factorial(n))/(factorial(i)*factorial(n-i))
    print(f"El resultado es {op}")



n: 4
i: 2
24
2
2
El resultado es 6.0
